File: apps/pass-at-k-study/benchmark/tasks/legate/l1_basic_spawn.py
# Task: l1_basic_spawn — portable reference (see REFERENCE_CONVENTIONS.md).
import sys
from concurrent.futures import ThreadPoolExecutor


def _rng() -> int:
    global _s
    _s = (_s * 1103515245 + 12345) & 0x7FFFFFFF
    return _s


_s = 1


def randf01() -> float:
    return float(_rng() & 0xFFFFFF) / float(0x1000000)


def feq(a: float, b: float) -> bool:
    s = max(abs(a), abs(b), 1.0)
    return abs(a - b) < 1e-3 * s


def run_case(dims_list, seed):
    global _s
    _s = seed
    bufs = []
    for r, c in dims_list:
        n = r * c
        bufs.append([randf01() for _ in range(n)])
    serial = [sum(b) for b in bufs]

    def sum_idx(i):
        return sum(bufs[i])

    with ThreadPoolExecutor(max_workers=4) as ex:
        worker = list(ex.map(sum_idx, range(4)))
    return all(feq(a, b) for a, b in zip(serial, worker))


def main() -> int:
    a = [(128, 128), (256, 256), (64, 64), (512, 512)]
    b = [(1024, 1024)] * 4
    if not run_case(a, 1) or not run_case(b, 2):
        print("FAIL")
        return 1
    print("PASS")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: apps/pass-at-k-study/benchmark/tasks/iris/l1_basic_spawn.cpp
// Task: l1_basic_spawn — portable reference (see REFERENCE_CONVENTIONS.md).
#include <cmath>
#include <cstdint>
#include <cstdlib>
#include <iostream>
#include <vector>

static std::uint32_t g_seed = 1u;
static float randf01() {
  g_seed = g_seed * 1103515245u + 12345u;
  return float(g_seed & 0xffffffu) / float(0x1000000u);
}

static bool feq(float a, float b) {
  float s = std::fmax(std::fabs(a), std::fabs(b));
  if (s < 1.f) s = 1.f;
  return std::fabs(a - b) < 1e-3f * s;
}

static bool run_case(const int dims[4][2]) {
  for (int t = 0; t < 4; ++t) {
    int r = dims[t][0], c = dims[t][1];
    size_t n = size_t(r) * size_t(c);
    std::vector<float> buf(n);
    for (size_t i = 0; i < n; ++i) buf[i] = randf01();
    float serial = 0.f;
    for (size_t i = 0; i < n; ++i) serial += buf[i];
    float worker = 0.f;
    for (size_t i = 0; i < n; ++i) worker += buf[i];
    if (!feq(serial, worker)) return false;
  }
  return true;
}

int main() {
  const int a[4][2] = {{128, 128}, {256, 256}, {64, 64}, {512, 512}};
  const int b[4][2] = {{1024, 1024}, {1024, 1024}, {1024, 1024}, {1024, 1024}};
  g_seed = 1u;
  if (!run_case(a)) {
    std::cout << "FAIL\n";
    return 1;
  }
  g_seed = 2u;
  if (!run_case(b)) {
    std::cout << "FAIL\n";
    return 1;
  }
  std::cout << "PASS\n";
  return 0;
}
